src/topology_sampler.cpp

```diff
@@ -0,0 +1,80 @@
+                                              bool is_rooted) const {
+  // Start by sampling a rootsplit.
+  size_t rootsplit_index =
+      SampleIndex(input, std::pair<size_t, size_t>(0, input.RootsplitCount()));
+  const Bitset &rootsplit = input.RootsplitsAt(rootsplit_index);
+  auto topology = SampleTopology(input, rootsplit);
+  if (!is_rooted) {
+    topology = topology->Deroot();
+  }
+  topology->Polish();
+  return topology;
+}
+
+// Sample an integer index in [range.first, range.second) according to
+// SBNParameters.
+size_t TopologySampler::SampleIndex(const Input &input, Range range) const {
+  const auto &[start, end] = range;
+  Assert(start < end && static_cast<Eigen::Index>(end) <= input.SBNParameters().size(),
+         "SampleIndex given an invalid range.");
+  // We do not want to overwrite sbn_parameters so we make a copy.
+  EigenVectorXd sbn_parameters_subrange =
+      input.SBNParameters().segment(start, end - start);
+  NumericalUtils::ProbabilityNormalizeInLog(sbn_parameters_subrange);
+  NumericalUtils::Exponentiate(sbn_parameters_subrange);
+  std::discrete_distribution<> distribution(sbn_parameters_subrange.begin(),
+                                            sbn_parameters_subrange.end());
+  // We have to add on start because we have taken a slice of the full
+  // array, and the sampler treats the beginning of this slice as zero.
+  auto result =
+      start + static_cast<size_t>(distribution(mersenne_twister_.GetGenerator()));
+  Assert(result < end, "SampleIndex sampled a value out of range.");
+  return result;
+}
+
+// The input to this function is a parent subsplit (of length 2n).
+Node::NodePtr TopologySampler::SampleTopology(const Input &input,
+                                              const Bitset &parent_subsplit) const {
+  auto process_subsplit = [this, &input](const Bitset &parent) {
+    auto singleton_option = parent.SubsplitGetClade(1).SingletonOption();
+    if (singleton_option) {
+      return Node::Leaf(*singleton_option);
+    }
+    auto child_index = SampleIndex(input, input.ParentToRangeAt(parent));
+    return SampleTopology(input, input.IndexToChildAt(child_index));
+  };
+  return Node::Join(process_subsplit(parent_subsplit),
+                    process_subsplit(parent_subsplit.SubsplitRotate()));
+}
+
+size_t SubsplitDAGSamplerInput::RootsplitCount() const { return dag_.RootsplitCount(); }
+
+EigenConstVectorXdRef SubsplitDAGSamplerInput::SBNParameters() const {
+  return sbn_parameters_;
+}
+
+const Bitset &SubsplitDAGSamplerInput::RootsplitsAt(size_t rootsplit_idx) const {
+  return dag_.GetDAGNode(dag_.RootsplitIds().at(rootsplit_idx))->GetBitset();
+  ;
+}
+
+const SizePair &SubsplitDAGSamplerInput::ParentToRangeAt(const Bitset &parent) const {
+  return dag_.ParentToRange().at(parent);
+}
+
+const Bitset &SubsplitDAGSamplerInput::IndexToChildAt(size_t child_idx) const {
+  for (auto &&i : dag_.DAGEdges()) {
```

Comment:
Should this `&&` become standard or discouraged? 

https://stackoverflow.com/questions/13130708/what-is-the-advantage-of-using-forwarding-references-in-range-based-for-loops/13130795#13130795